Source/WebCore/rendering/RenderLayerCompositor.cpp

```diff
@@ -1165,7 +1164,20 @@ bool RenderLayerCompositor::updateCompositingLayers(CompositingUpdateType update
     return true;
 }
 
-void RenderLayerCompositor::computeCompositingRequirements(RenderLayer* ancestorLayer, RenderLayer& layer, LayerOverlapMap& overlapMap, CompositingState& compositingState, BackingSharingState& backingSharingState, bool& descendantHas3DTransform)
+// Unchanged leaf compositing layers that clip their descendants can skip descendant
+// traversal, since their descendants can't contribute any new overlap to the map.
+static bool canSkipComputeCompositingRequirementsForSubtree(RenderLayer& layer, bool willBeComposited)
+{
+    if (layer.needsCompositingRequirementsTraversal() || layer.hasDescendantNeedingCompositingRequirementsTraversal())
+        return false;
+
+    if (!layer.isComposited() || !willBeComposited || layer.hasCompositingDescendant())
```

Comment:
I'm not 100% convinced the `layer.hasCompositingDescendant()` check is necessary. It seems likely that we could skip over subtrees with compositing layers inside them, I don't think overlap changes from before the current layer can affect descendants (which must not extend outside of this layers overlap bounds due to the clipping).

I also didn't make this a check for stacking context. I think the logic is correct for the paint-order children of this layer. Stacking contexts just have more paint-order children that would be affected by the optimisation.

Reply:
I worry about mutations on ancestors or earlier siblings that require us to traverse into the "unchanged" subtree to do things like scrolling tree node registration.

I think it would be safer to constrain this change to stacking contexts for now.

Reply:
If we do have this sort of problem, then how does a stacking context avoid it?